features/timeseries/src/main/java/org/opennms/netmgt/timeseries/integration/persistence/MetaTagDataLoader.java

```diff
@@ -0,0 +1,286 @@
+ *     http://www.opennms.com/
+ *******************************************************************************/
+
+package org.opennms.netmgt.timeseries.integration.persistence;
+
+import java.net.InetAddress;
+import java.net.UnknownHostException;
+import java.util.Date;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.concurrent.atomic.AtomicInteger;
+
+import javax.inject.Inject;
+
+import org.opennms.netmgt.collection.api.CollectionResource;
+import org.opennms.netmgt.dao.api.NodeDao;
+import org.opennms.netmgt.dao.api.SessionUtils;
+import org.opennms.netmgt.model.OnmsAssetRecord;
+import org.opennms.netmgt.model.OnmsCategory;
+import org.opennms.netmgt.model.OnmsNode;
+import org.opennms.netmgt.model.ResourceTypeUtils;
+import org.opennms.netmgt.model.monitoringLocations.OnmsMonitoringLocation;
+import org.opennms.netmgt.timeseries.integration.MetaTagConfiguration;
+import org.opennms.netmgt.timeseries.integration.MetaTagConfiguration.AssetTagKey;
+import org.opennms.netmgt.timeseries.integration.MetaTagConfiguration.MetaTagKey;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.common.base.Strings;
+import com.google.common.cache.CacheLoader;
+import com.google.common.collect.Maps;
+
+public class MetaTagDataLoader extends CacheLoader<CollectionResource, Map<String, String>> {
+
+    private static final Logger LOG = LoggerFactory.getLogger(MetaTagDataLoader.class);
+
+    private final NodeDao nodeDao;
+    private final SessionUtils sessionUtils;
+    private MetaTagConfiguration config;
+
+    @Inject
+    public MetaTagDataLoader(final NodeDao nodeDao, final SessionUtils sessionUtils) {
+        this.nodeDao = nodeDao;
+        this.sessionUtils = sessionUtils;
+        setConfig(new MetaTagConfiguration(Maps.fromProperties(System.getProperties())));
+    }
+
+    public void setConfig(final MetaTagConfiguration config) {
+        this.config = config;
+    }
+
+    public Map<String, String> load(final CollectionResource resource) {
+        return sessionUtils.withReadOnlyTransaction(() -> {
+            final Map<String, String> tags = new HashMap<>();
+            addTag(tags, MetaTagKey.resourceLabel, Optional.ofNullable(resource.getInterfaceLabel()).orElse(null));
+
+            if (resource.getResourceTypeName().equals(CollectionResource.RESOURCE_TYPE_NODE)) {
+                String nodeCriteria = getNodeCriteriaFromResource(resource);
+                mapNode(tags, nodeCriteria);
+            } else if (resource.getResourceTypeName().equals(CollectionResource.RESOURCE_TYPE_IF)) {
+                String nodeCriteria = getNodeCriteriaFromResource(resource);
+                if (!Strings.isNullOrEmpty(nodeCriteria)) {
+                    mapNode(tags, nodeCriteria);
+                }
+                addTag(tags, MetaTagKey.ifDescr, resource.getInterfaceLabel()); // TODO: Patrick: is that correct?
```

Comment:
@j-white ?

Reply:
We'll need to extract the ifIndex from the resource to be able to get everything we need.